src/derivative_utils.jl

```diff
@@ -312,19 +325,24 @@ function calc_W!(integrator, cache::OrdinaryDiffEqMutableCache, dtgamma, repeat_
     # skip calculation of inv(W) if step is repeated
     (!repeat_step && W_transform) ? f.invW_t(W, uprev, p, dtgamma, t) : f.invW(W, uprev, p, dtgamma, t) # W == inverse W
     is_compos && calc_J!(integrator, cache, true)
-  elseif DiffEqBase.has_jac(f) && f.jac_prototype !== nothing
+  elseif (freshdt = isnewton ? cache.nlsolver.cache.freshdt : dt; # TODO: RosW
```

Comment:
why is this all in the elseif?

Reply:
They will also be used in the `else` case.

Reply:
still kind of kludgy to read